### PR TITLE
debugger: Fix install detection for debugpy

### LINKED ISSUE #31383: Debugger: <It does not work with a virtual environment.>
### Summary

On Zed Preview 0.188.2

.zed/debug.json:

[
  {
    "adapter": "Debugpy",
    "label": "Bebugger",
    "request": "launch",
    "subProcess": true,
    "redirectOutput": true,
    "program": "main.py",
    "args": [],
    "stopOnEntry": false,
    "cwd": "$ZED_WORKTREE_ROOT/boriy",
    "python": "$ZED_WORKTREE_ROOT/boriy/.venv/bin/python"
  }
]
The debugger starts, but when you try to make a request to the application to check the contents of the variables, the debugger shuts down and stops with an error.

Error:
2025-05-25 16:43:36 +0300] [5062] [INFO] Starting gunicorn 23.0.0
[2025-05-25 16:43:36 +0300] [5062] [INFO] Listening at: http://0.0.0.0:5555 (5062)
[2025-05-25 16:43:36 +0300] [5062] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2025-05-25 16:43:36 +0300] [5065] [INFO] Booting worker with pid: 5065
[2025-05-25 16:43:36 +0300] [5065] [INFO] Started server process [5065]
[2025-05-25 16:43:36 +0300] [5065] [INFO] Waiting for application startup.
[2025-05-25 16:43:36 +0300] [5065] [INFO] Application startup complete.
objc[5065]: +[NSMutableString initialize] may have been in progress in another thread when fork() was called.
objc[5065]: +[NSMutableString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[2025-05-25 16:43:44 +0300] [5062] [ERROR] Worker (pid:5065) was sent SIGABRT!

P.S.
A FastAPI Python application with a virtual environment!

Image

Image

### Zed Version and System Specs

Zed: v0.188.2 (Zed Preview) 
OS: macOS 15.5.0
Memory: 8 GiB
Architecture: aarch64

### PR DESCRIPTION
**Pull request by raphaelluethy**
Closes: #31383 

Based on my report on discord when chatting with Anthony and Remco: https://discord.com/channels/869392257814519848/1375129714645012530

Root Cause: Zed was incorrectly trying to execute a directory path instead of properly invoking the debugpy module when debugpy was installed via package managers (pip, conda, etc.) rather than downloaded from GitHub releases.

Solution:

- Automatic Detection: Zed now automatically detects whether debugpy is installed via pip/conda or downloaded from GitHub
- Correct Invocation: For pip-installed debugpy, Zed now uses python -m debugpy.adapter instead of trying to execute file paths
- Added a `installed_in_venv` flag to differentiate the setup properly
- Backward Compatibility: GitHub-downloaded debugpy releases continue to work as before
- Enhanced Logging: Added logging to show which debugpy installation method is being used (I had to verify it somehow)

I verified with the following setups (can be confirmed with the debug logs):
- `conda` with installed debugpy, went to installed instance
- `uv` with installed debugpy, went to installed instance
- `uv` without installed debugpy, went to github releases
- Homebrew global python install, went to github releases

Release Notes:

- Fix issue where debugpy from different environments won't load as intended

### CODE REVIEW

File: crates/dap_adapters/src/python.rs [hunk 1]
```diff
@@ -128,16 +115,75 @@ impl PythonDebugAdapter {
...
         };
 
-        Ok(DebugAdapterBinary {
-            command: python_path.context("failed to find binary path for Python")?,
-            arguments: vec![
+        let python_command = python_path.context("failed to find binary path for Python")?;
+        log::info!("Using Python executable: {}", python_command);
+
+        let arguments = if let Some(user_installed_path) = user_installed_path {
+            if user_installed_path
```
Comment (RemcoSmitsDev) [inline crates/dap_adapters/src/python.rs:122]: I think this assumption is not right, because I have debugpy locally installed without it being installed via a package manager. Maybe passing in a bool flag that we know it's installed via package manager is better.

File: crates/dap_adapters/src/python.rs [hunk 2]
```diff
@@ -128,16 +115,75 @@ impl PythonDebugAdapter {
...
+            {
+                log::info!(
+                    "Using pip-installed debugpy from: {}",
+                    user_installed_path.display()
+                );
+                vec![
+                    "-m".to_string(),
+                    "debugpy.adapter".to_string(),
+                    "--host".to_string(),
+                    host.to_string(),
```
Comment (RemcoSmitsDev) [inline crates/dap_adapters/src/python.rs:135]: Just a nit, but can have the same style as we do for the fallback.
`format!("--port={}", port)`

**Review by RemcoSmitsDev — CHANGES_REQUESTED** [review]

File: crates/dap_adapters/src/python.rs (hunk 1 above)
Comment (raphaelluethy) [reply crates/dap_adapters/src/python.rs:122]: So would you like this to be a setting? Would this also mean that there is a need for a path option for the custom install or do we assume its located in .bin?

Comment (RemcoSmitsDev) [reply crates/dap_adapters/src/python.rs:122]: So my idea was instead of passing in the venv debugpy path (..venv/bin/debugpy) inside the `get_installed_binary` method we could pass in a boolean when we detected that debugpy is installed inside the user's venv. So we could add a second parameter to the `get_installed_binary` method.

So instead of checking `if let some(...) = user_installed_path` we should do `if installed_in_venv` and do the `debugpy.adapter` code path. And if it's false we can check for the old code path with the user_installed_path variable and if that does not exist we can do the download from GitHub code.

Comment (raphaelluethy) [reply crates/dap_adapters/src/python.rs]: I added the argument, please have a look :)

Comment (RemcoSmitsDev) [reply crates/dap_adapters/src/python.rs]: Sorry, but I wrote down the wrong order. We should do the `user_installed_path` check first so you can always override it.

File: crates/dap_adapters/src/python.rs [hunk 3]
```diff
@@ -27,6 +28,45 @@ impl PythonDebugAdapter {
...
+                github_path
+                    .join(Self::ADAPTER_PATH)
+                    .to_string_lossy()
+                    .to_string(),
+                format!("--port={}", port),
+                format!("--host={}", host),
+            ]
+        } else {
+            // This shouldn't happen in practice, but provide a fallback
+            vec![format!("--port={}", port), format!("--host={}", host)]
```
Comment (raphaelluethy) [inline crates/dap_adapters/src/python.rs:66]: We might need to handle this error since this should never happen, or do we just leave it as is?

Comment (RemcoSmitsDev) [reply crates/dap_adapters/src/python.rs:66]: I think we can just error out here

File: crates/dap_adapters/src/python.rs [hunk 4]
```diff
@@ -27,6 +28,60 @@ impl PythonDebugAdapter {
...
+    ) -> Result<Vec<String>> {
+        if installed_in_venv {
+            log::debug!("Using venv-installed debugpy");
+            Ok(vec![
+                "-m".to_string(),
+                "debugpy.adapter".to_string(),
+                format!("--host={}", host),
+                format!("--port={}", port),
+            ])
+        } else if let Some(user_installed_path) = user_installed_path {
```
Comment (Anthony-Eid) [inline crates/dap_adapters/src/python.rs:46]: Because `user_installed_path` is a setting I take it should take precedence over a virtual environment installation. Do you have strong feelings about checking for venv first?

Edit: 

Based on your test and the callers of this function `user_installed_path` does take precedence over `installed_in_venv`. However, I still think you should change the order of the if statements to increase readability and in case this function is called from someone else in the future.

File: crates/dap_adapters/src/python.rs [hunk 5]
```diff
@@ -572,12 +636,17 @@ impl DebugAdapter for PythonDebugAdapter {
             if let Some(path) = Path::new(&toolchain.path.to_string()).parent() {
                 let debugpy_path = path.join("debugpy");
                 if smol::fs::metadata(&debugpy_path).await.is_ok() {
```
Comment (Anthony-Eid) [inline crates/dap_adapters/src/python.rs:638]:
```suggestion
                if delegate.fs().is_file(&debugpy_path).await {
```

We have a `Fs` trait that is usually accessible from `AppState`, which is on `Workspace` among other types. 

If you need to do file system operations, it's a good idea to see if the `Fs` trait has a method that does what you need. It's more idiomatic and more readable (in this case).

File: crates/dap_adapters/src/python.rs [hunk 6]
```diff
@@ -591,7 +660,52 @@ impl DebugAdapter for PythonDebugAdapter {
...
+
+        assert_eq!(venv_args[0], "-m");
+        assert_eq!(venv_args[1], "debugpy.adapter");
+        assert_eq!(venv_args[2], "--host=127.0.0.1");
+        assert_eq!(venv_args[3], "--port=5678");
+
+        // Note: Case 3 (GitHub-downloaded debugpy) is now handled internally
+        // by the generate_debugpy_arguments function when neither user_installed_path
+        // nor installed_in_venv are provided, so we don't test it separately here
+        // as it would require setting up the actual file system structure.
```
Comment (Anthony-Eid) [inline crates/dap_adapters/src/python.rs:704]: Just for future noticed we have a `FakeFs` struct that's used for tests. I don't expected you to add that edge case to merge this PR though. I don't think it would be a good idea that would rely on a GitHub API to pass or mocking the behavior we would be testing.

**Comment by Anthony-Eid** [comment]
@raphaelluethy Thank you so much for fixing this bug! This PR is looking really good so far, but there are some small changes I think we could make to further improve it. I can push this PR across the finish line on Tuesday if you don't have time to make the changes.  Otherwise, I'll be happy to merge it once the feedback from me and Remco has been addressed.

@RemcoSmitsDev Thank you for taking the initiative to start a PR review! You did a fantastic job, and I really appreciate all your help.

File: crates/dap_adapters/src/python.rs [hunk 7]
```diff
@@ -27,6 +28,45 @@ impl PythonDebugAdapter {
...
+            vec![
+                github_path
+                    .join(Self::ADAPTER_PATH)
+                    .to_string_lossy()
+                    .to_string(),
+                format!("--port={}", port),
+                format!("--host={}", host),
+            ]
+        } else {
+            // This shouldn't happen in practice, but provide a fallback
```
Comment (Anthony-Eid) [inline crates/dap_adapters/src/python.rs]:
```suggestion
            debug_panic!("Failed to download/get debugpy from venv, user installed path, and github")
```

This will only panic in debug builds and log an error message in production builds. I like using these panics a lot because it's a great way to find bugs and check that the code's logic is correct.

**Comment by raphaelluethy** [comment]
@Anthony-Eid Thank you, if there is more specific/advanced stuff I would be glad if you could fix it, I think the issues from @RemcoSmitsDev should be mostly addressed.

If there are open issues/comments please add them and I will they if I can resolve them.

**Review by Anthony-Eid — CHANGES_REQUESTED** [review]

File: crates/dap_adapters/src/python.rs (hunk 4 above)
Comment (raphaelluethy) [reply crates/dap_adapters/src/python.rs:46]: I have the venv one first in order to reduce the work being done with checking the options. I have changed it

File: crates/dap_adapters/src/python.rs (hunk 5 above)
Comment (raphaelluethy) [reply crates/dap_adapters/src/python.rs:638]: done 👍

File: crates/dap_adapters/src/python.rs (hunk 6 above)
Comment (raphaelluethy) [reply crates/dap_adapters/src/python.rs:704]: I updated the comment 👍

File: crates/dap_adapters/src/python.rs (hunk 7 above)
Comment (raphaelluethy) [reply crates/dap_adapters/src/python.rs]: This not necessary anymore since the fallback is the github download. I changed this when extracting the if statement to the new function.

**Review by Anthony-Eid — APPROVED** [review]
